src/kernels/layer.py

```diff
@@ -119,96 +110,125 @@ def register_kernel_mapping(
                 device_repo[new_device] = new_repo
 
 
-def replace_kernel_forward_from_hub(cls, layer_name: str, *, use_fallback: bool = True):
+def replace_kernel_forward_from_hub(
+    cls,
+    layer_name: str,
+    *,
+):
     """
-    Replace the forward function of a layer using a layer from the kernel hub.
-    This function monkeypatches a layer, replacing the `forward` method
-    of the layer with that of a layer from the hub. The replacement is done
-    when a layer matching `layer_name` and device type is registered through
-    `register_layer_mapping`. The device type is inferred from the first
-    argument to `forward`.
+    Decorator that prepares a layer class to use a kernel from the Hugging Face Hub.
+    
+    This decorator stores the layer name and original forward method, which will be used
+    by the kernelize function to replace the forward implementation with the appropriate
+    kernel from the hub.
+    
+    Args:
+        cls: The layer class to decorate
+        layer_name: The name of the layer to use for kernel lookup
     """
+    cls.kernel_layer_name = layer_name
+    cls.fallback_forward = cls.forward
 
-    fallback_forward = cls.forward
 
-    cached_layer: Dict[LayerRepository, nn.Module] = {}
+cached_layer: ContextVar[Dict[LayerRepository, nn.Module]] = ContextVar("cached_layer", default={})
 
-    def forward(self, x, *args, **kwargs):
-        if _DISABLE_KERNEL_MAPPING:
-            return fallback_forward(self, x, *args, **kwargs)
+def kernelize(
+    model: nn.Module,
+    device: Device = Device(type="cuda"),
```

Comment:
I think it would be nicer if this was:

```suggestion
    device: Optional[Device] = None,
```

Behavior: if the user passes a device type, use it. If the user does not pass a device type, get it from a (probably the first, since individual layers may not have parameters) parameter of the model. Similarly, I would add an extra argument:

```
    needs_backward: Optional[bool] = None,
```

If the user sets a bool, then do as they wish. Otherwise, use the logic that is already there below. This has the benefit that if a model is in `eval()`, the user could still kernelize it with kernels that support backward in case they want to move to `train()` later.

Then, with these new options, I would make everything after `model` kwarg-only. Positional arguments lead to a lot of bugs/failures.